lib/Sema/CSSimplify.cpp

```diff
@@ -3425,18 +3425,36 @@ ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
       if (!anchor)
         return getTypeMatchFailure(argumentLocator);
 
-      // If there are missing arguments, let's add them
-      // using parameter as a template.
-      if (diff < 0) {
-        if (fixMissingArguments(*this, anchor, func1Params, func2Params,
-                                abs(diff), loc))
+      // The param diff is in a function type coercion context
+      //
+      // func fn(_: Int) {}
+      // let i: Int = 0
+      // (fn as (Int, Int) -> Void)(i, i)
+      //
+      // Since we are not in a function argument application, so simply record
+      // a function type mismatch instead of an argument fix.
+      // Except for when a closure is a subexpr because missing and extraneous
+      // arg fix can properly handle closure diagnostic.
```

Comment:
Nit: I think we should rework this section a bit to explain the reason why the closure is special - because it's possible to suggest source code changes for closures to remove or add arguments (which is handled by missing/extraneous arguments fix) unlike function references.

Reply:
Yes, just updated =]